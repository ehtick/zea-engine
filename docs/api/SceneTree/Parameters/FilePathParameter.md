<a name="FilePathParameter"></a>

### FilePathParameter 
Represents a specific type of parameter, that only stores file data values.

**Events**
* **valueChanged:** Triggered when setting file's URL.
* **fileUpdated:** Triggered when parameter's value is updated.


**Extends**: <code>[Parameter](api/SceneTree\Parameters\Parameter.md)</code>  

* [FilePathParameter ⇐ <code>Parameter</code>](#FilePathParameter)
    * [new FilePathParameter(name, exts)](#new-FilePathParameter)
    * [getFilepath() ⇒ <code>string</code>](#getFilepath)
    * [setFilepath(filePath)](#setFilepath)
    * [getFilename() ⇒ <code>string</code>](#getFilename)
    * [getExt() ⇒ <code>string</code>](#getExt)
    * [getStem() ⇒ <code>string</code>](#getStem)
    * [getFileDesc() ⇒ <code>object</code>](#getFileDesc)
    * [getFile() ⇒ <code>object</code>](#getFile)
    * [setUrl(url, name)](#setUrl)
    * [getUrl() ⇒ <code>string</code>](#getUrl)
    * [setValue(value)](#setValue)
    * [toJSON(context) ⇒ <code>object</code>](#toJSON)
    * [fromJSON(j, context)](#fromJSON)
    * [clone()](#clone)

<a name="new_FilePathParameter_new"></a>

### new FilePathParameter
Create a file path parameter.


| Param | Type | Description |
| --- | --- | --- |
| name | <code>string</code> | The name of the file path parameter. |
| exts | <code>string</code> | The exts value. |

<a name="FilePathParameter+getFilepath"></a>

### getFilepath
Returns complete file path.


**Returns**: <code>string</code> - - The return value.  
<a name="FilePathParameter+setFilepath"></a>

### setFilepath
Resolves resourceId using the specified path and sets its value to the parameter.



| Param | Type | Description |
| --- | --- | --- |
| filePath | <code>string</code> | The filePath value. |

<a name="FilePathParameter+getFilename"></a>

### getFilename
Returns parameter's file name


**Returns**: <code>string</code> - - The return value.  
<a name="FilePathParameter+getExt"></a>

### getExt
Returns parameter's file extension


**Returns**: <code>string</code> - - The return value.  
<a name="FilePathParameter+getStem"></a>

### getStem
Returns parameter's file name without extension


**Returns**: <code>string</code> - - The return value.  
<a name="FilePathParameter+getFileDesc"></a>

### getFileDesc
Returns file object, which contains the url, resourceId and the name.


**Returns**: <code>object</code> - - The return value.  
<a name="FilePathParameter+getFile"></a>

### getFile
Returns file object, which contains the url, resourceId and the name.


**Returns**: <code>object</code> - - The return value.  
<a name="FilePathParameter+setUrl"></a>

### setUrl
Sets file data.



| Param | Type | Description |
| --- | --- | --- |
| url | <code>string</code> | the url value of the |
| name | <code>string</code> | (optional) the name of the file that the Url points to. |

<a name="FilePathParameter+getUrl"></a>

### getUrl
Returns the file url string.


**Returns**: <code>string</code> - - The return value.  
<a name="FilePathParameter+setValue"></a>

### setValue
Sets file parameter value



| Param | Type | Description |
| --- | --- | --- |
| value | <code>string</code> | The value param. |

<a name="FilePathParameter+toJSON"></a>

### toJSON
The toJSON method encodes this type as a json object for persistence.


**Returns**: <code>object</code> - - Returns the json object.  

| Param | Type | Description |
| --- | --- | --- |
| context | <code>object</code> | The context value. |

<a name="FilePathParameter+fromJSON"></a>

### fromJSON
The fromJSON method decodes a json object for this type.



| Param | Type | Description |
| --- | --- | --- |
| j | <code>object</code> | The json object this item must decode. |
| context | <code>object</code> | The context value. |

<a name="FilePathParameter+clone"></a>

### clone
The clone method constructs a new file path parameter,
copies its values from this parameter and returns it.


**Returns**: [<code>FilePathParameter</code>](#FilePathParameter) - - Returns a new cloned file path parameter.  


### [Class Tests](api/SceneTree\Parameters/FilePathParameter.test)