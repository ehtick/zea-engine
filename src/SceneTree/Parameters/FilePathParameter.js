import { ParamFlags, Parameter } from './Parameter.js'
import { resourceLoader } from '../ResourceLoader.js'

/**
 * Represents a specific type of parameter, that only stores file data values.
 *
 * **Events**
 * * **valueChanged:** Triggered when setting file's URL.
 * * **fileUpdated:** Triggered when parameter's value is updated.
 *
 * @extends Parameter
 */
class FilePathParameter extends Parameter {
  /**
   * Create a file path parameter.
   *
   * @param {string} name - The name of the file path parameter.
   * @param {string} exts - The exts value.
   */
  constructor(name, exts) {
    super(name, '', 'FilePath')
    if (exts) this.setSupportedExts(exts)
  }

  /**
   * Sets supported extensions, if this supports more than one type of files, separate them with regex or(|).
   *
   * i.e.: jpg|png|gif
   *
   * @param {string} exts - The exts value.
   */
  setSupportedExts(exts) {
    // Note: supported Extensions should be in the format ext1|exts2|ext3
    this.__reextensions = new RegExp('\\.(' + exts + ')$', 'i')
  }

  /**
   * Returns complete file path.
   *
   * @return {string} - The return value.
   */
  getFilepath() {
    if (this.__file) {
      return resourceLoader.getFilepath(this.__file.id)
    }

    return ''
  }

  /**
   * Resolves resourceId using the specified path and sets its value to the parameter.
   *
   * @param {string} filePath - The filePath value.
   */
  setFilepath(filePath) {
    const resourceId = resourceLoader.resolveFilePathToId(filePath)
    if (!resourceId) {
      console.warn('Resource unavailable:' + filePath)
      return
    }
    this.setValue(resourceId)
  }

  /**
   * Returns parameter's file name
   *
   * @return {string} - The return value.
   */
  getFilename() {
    if (this.__file) {
      return this.__file.name
    }
  }

  /**
   * Returns parameter's file extension
   *
   * @return {string} - The return value.
   */
  getExt() {
    const filename = this.getFilename()
    const suffixSt = filename.lastIndexOf('.')
    if (suffixSt != -1) return filename.substring(suffixSt).toLowerCase()
  }

  /**
   * Returns parameter's file name without extension
   *
   * @return {string} - The return value.
   */
  getStem() {
    const filename = this.getFilename()
    if (filename) {
      const parts = filename.split('.')
      if (parts.length == 2) return parts[0]
      else return filename
    }
  }

  /**
   * Returns parent folder for of current parameter file.
   *
   * @return {object} - The return value.
   */
  getFileFolder() {
    if (this.__file) {
      if (this.__file.parent) return resourceLoader.getFile(this.__file.parent)
      return resourceLoader.getRootFolder()
    }
  }

  /**
   * Returns parent folder for of current parameter file.
   *
   * @return {string} - The return value.
   */
  getFileFolderPath() {
    const filePath = this.getFilepath()
    if (filePath) {
      return filePath.substring(0, filePath.lastIndexOf('/')) + '/'
    }
  }

  /**
   * Returns file object, which contains the url, resourceId and the name.
   *
   * @return {object} - The return value.
   */
  getFile() {
    return this.__file
  }

  /**
   * The getFileDesc method.
   * @return {object} - The return value.
   */
  getFileDesc() {
    console.warn('@todo-review')
    // Can we settle on a convention?
    // console.warn("Deprecated method: 'getFileDesc'. Please use 'getFile'")
    return this.__file
  }

  /**
   * Sets file data.
   *
   * @param {string} url - the url value of the
   * @param {string} name - (optional) the name of the file that the Url points to.
   */
  setUrl(url, name) {
    const parts = url.split('/')
    if (!name) name = parts[parts.length - 1]

    this.__value = name
    this.__file = {
      id: url,
      name,
      url,
    }

    this.__flags |= ParamFlags.USER_EDITED
    this.emit('valueChanged', { mode: ParamFlags.USER_EDITED })
  }

  /**
   * Returns the file url string.
   *
   * @return {string} - The return value.
   */
  getUrl() {
    return this.__file ? this.__file.url : undefined
  }

  /**
   * The setDirty method.
   * @private
   * @param {function} cleanerFn - The cleanerFn value.
   */
  setDirty(cleanerFn) {
    console.warn('@todo-review')
    throw new Error('Cannot drive a filepath param from an operator')
  }

  /**
   * Sets file parameter value receiving its resource id.
   *
   * @param {string} value - The value param.
   * @return {boolean} - The return value.
   */
  setValue(value) {
    // 0 == normal set. 1 = changed via cleaner fn, 2 = change by loading/cloning code.
    if (value == undefined) {
      throw new Error('Invalid value for setValue.')
    }
    if (value.indexOf('.') > 0) {
      console.warn('Deprecation warning for setValue. setValue should now only take a file id, not a path.')
      return this.setFilepath(value)
    }
    // Note: equality tests only work on simple types.
    // Important here because file changes cause reloads..
    if (value == this.__value) {
      return
    }

    // Note: the file path is selected by using the file browser
    // For now it can return an absolute path(within the project)
    // and we convert to relative when we save.
    const resourceId = value
    if (!resourceLoader.resourceAvailable(resourceId)) {
      console.warn('Resource unavailable:' + resourceId)
      return
    }

    const file = resourceLoader.getFile(resourceId)
    if (this.__reextensions && !this.__reextensions.test(file.name)) {
      console.warn('Unsupported file type:' + file.name)
      return false
    }

    this.__value = value
    this.__file = file

    resourceLoader.on('fileUpdated', (event) => {
      if (event.fileId == this.__value) {
        this.__file = resourceLoader.getFile(this.__value)
        this.emit('fileUpdated', event)
      }
    })

    this.__flags |= ParamFlags.USER_EDITED
    this.emit('valueChanged', { mode: ParamFlags.USER_EDITED })
  }
  // ////////////////////////////////////////
  // Persistence

  /**
   * The toJSON method encodes this type as a json object for persistence.
   *
   * @param {object} context - The context value.
   * @param {number} flags - The flags value.
   * @return {object} - Returns the json object.
   */
  toJSON(context, flags) {
    if ((this.__flags & ParamFlags.USER_EDITED) == 0) return
    const j = {}
    if (this.__file) {
      j.value = this.__file.id
      // For cases where the file ID changed.
      // e.g. if a file was deleted from the system, and
      // then re-added
      j.filepath = resourceLoader.getFilepath(this.__file.id)
    }
    return j
  }

  /**
   * The fromJSON method decodes a json object for this type.
   *
   * @param {object} j - The json object this item must decode.
   * @param {object} context - The context value.
   * @param {number} flags - The flags value.
   */
  fromJSON(j, context, flags) {
    if (j.value) {
      if (j.value.indexOf('.') > 0) {
        this.__value = j.value
      } else {
        if (resourceLoader.resourceAvailable(j.value)) {
          this.__value = j.value
          this.__flags |= ParamFlags.USER_EDITED
        }
      }
    } else if (j.filepath) {
      const resourceId = resourceLoader.resolveFilePathToId(j.filepath)
      if (!resourceId) {
        console.warn('Resource unavailable:' + j.filepath)
      } else {
        this.__value = resourceId
      }
    }
  }

  // ////////////////////////////////////////
  // Clone and Destroy

  /**
   * The clone method constructs a new file path parameter,
   * copies its values from this parameter and returns it.
   *
   * @param {number} flags - The flags value.
   * @return {FilePathParameter} - Returns a new cloned file path parameter.
   */
  clone(flags) {
    const clonedParam = new FilePathParameter(this.__name)
    clonedParam.__file = this.__file
    return clonedParam
  }

  /**
   * The destroy is called by the system to cause explicit resources cleanup.
   * Users should never need to call this method directly.
   */
  destroy() {
    super.destroy()
  }
}

export { FilePathParameter }
