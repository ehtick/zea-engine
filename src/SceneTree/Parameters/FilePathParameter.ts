/* eslint-disable @typescript-eslint/no-explicit-any */
/* eslint-disable @typescript-eslint/no-unused-vars */
import { Registry } from '../../Registry'
import { Parameter } from './Parameter'
import { resourceLoader } from '../resourceLoader'

/**
 * Represents a specific type of parameter, that only stores file data values.
 *
 * **Events**
 * * **valueChanged:** Triggered when setting file's URL.
 * * **fileUpdated:** Triggered when parameter's value is updated.
 *
 * @extends Parameter
 */
class FilePathParameter extends Parameter<string> {
  /**
   * Create a file path parameter.
   *
   * @param {string} name - The name of the file path parameter.
   * @param {string} exts - The exts value.
   */
  constructor(name: string = '') {
    super(name, '', 'FilePath')
  }

  /**
   * Returns complete file path.
   *
   * @return {string} - The return value.
   */
  getFilepath(): string {
    if (this.value) {
      return (resourceLoader as any).getFilepath(this.value)
    }

    return ''
  }

  /**
   * Resolves resourceId using the specified path and sets its value to the parameter.
   *
   * @param {string} filePath - The filePath value.
   */
  setFilepath(filePath: string): void {
    this.setValue(resourceLoader.resolveFileId(filePath))
  }

  /**
   * Returns parameter's file name
   *
   * @return {string} - The return value.
   */
  getFilename(): string {
    if (!this.value) throw 'No file value'

    return resourceLoader.resolveFilename(this.value)
  }

  /**
   * Returns parameter's file extension
   *
   * @return {string} - The return value.
   */
  getExt(): string | undefined {
    const filename = this.getFilename()
    const suffixSt = filename.lastIndexOf('.')
    if (suffixSt != -1) return filename.substring(suffixSt).toLowerCase()
  }

  /**
   * Returns parameter's file name without extension
   *
   * @return {string} - The return value.
   */
  getStem(): string | null{
    const filename = this.getFilename()
    if (filename) {
      const parts = filename.split('.')
      if (parts.length == 2) return parts[0]
      else return filename
    }
    return null
  }

  /**
   * Returns file object, which contains the url, resourceId and the name.
   *
   * @return {Record<string, string | undefined>} - The return value.
   */
  getFileDesc(): Record<string, string | undefined> {
    return this.getFile()
  }

  /**
   * Returns file object, which contains the url, resourceId and the name.
   *
   * @return {Record<string, string | undefined>} - The return value.
   */
  getFile(): Record<string, string | undefined> {
    return { id: this.value, url: this.getUrl(), name: this.getFilename() }
  }

  /**
   * Sets file data.
   *
   * @param {string} url - the url value of the
   * @param {string} name - (optional) the name of the file that the Url points to.
   */
  setUrl(url: string, name: string): void {
    this.setValue(resourceLoader.resolveFileId(url))
  }

  /**
   * Returns the file url string.
   *
   * @return {string} - The return value.
   */
  getUrl(): string {
    if (!this.value) throw 'No file value'
    return resourceLoader.resolveURL(this.value)
  }

  /**
   * Sets file parameter value
   *
   * @param {string} value - The value param.
   */
  setValue(value: string): void {
    if (!value) {
      throw new Error('Invalid value for setValue.')
    }

    // Important here because file changes cause reloads.
    // Note: equality tests only work on simple types.
    if (value == this.value) {
      return
    }

    this.value = value

    this.emit('valueChanged', {})
  }
  // ////////////////////////////////////////
  // Persistence

  /**
   * The toJSON method encodes this type as a json object for persistence.
   *
   * @param {Record<string, any>} [context] - The context value.
   * @return {Record<string, unknown>} - Returns the json object.
   */
  toJSON(context?: Record<string, any>): Record<string, unknown> {
    return { value: this.value }
  }

  /**
   * The fromJSON method decodes a json object for this type.
   *
   * @param {Record<string, unknown>} j - The json object this item must decode.
   * @param {Record<string, any>} [context] - The context value.
   */
  fromJSON(j: Record<string, unknown>, context?: Record<string, any>): void {
    if (j.value) {
      this.value = j.value as string
    }
  }

  // ////////////////////////////////////////
  // Clone and Destroy

  /**
   * The clone method constructs a new file path parameter,
   * copies its values from this parameter and returns it.
   *
   * @return {FilePathParameter} - Returns a new cloned file path parameter.
   */
  clone(): FilePathParameter {
    const clone = new FilePathParameter(this.name)
    if (this.value) clone.setValue(this.value)

    return clone
  }
}

Registry.register('FilePathParameter', FilePathParameter)

export { FilePathParameter }
