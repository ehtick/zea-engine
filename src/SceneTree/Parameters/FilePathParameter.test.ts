import { FilePathParameter } from './FilePathParameter'

describe('FilePathParameter', () => {
  it('Has an initial value.', () => {
    const param = new FilePathParameter()

    expect(param.value).toBe('')
  })

  it('Has the correct value type.', () => {
    const param = new FilePathParameter()

    expect(param.getDataType()).toEqual('FilePath')
  })

  it("Can set it's value.", () => {
    const param = new FilePathParameter()
    param.value = 'foo'

    expect(param.value).toBe('foo')
  })

  it('Can be saved to JSON (serialization).', () => {
    const param = new FilePathParameter()
    param.value = 'foo'

    expect(JSON.stringify(param.toJSON())).toMatchSnapshot()
  })

  it('Can be loaded from JSON (deserialization).', () => {
    const param = new FilePathParameter()
    param.fromJSON({ value: 'foo' })

    expect(param.value).toBe('foo')
  })

  it('Can be cloned.', () => {
    const parameter = new FilePathParameter('TestParameter')
    parameter.value = 'foo'

    const clone = parameter.clone()

    expect(clone.toJSON()).toMatchSnapshot()
  })
})
