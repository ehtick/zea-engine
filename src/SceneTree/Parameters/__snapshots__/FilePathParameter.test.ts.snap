// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FilePathParameter Can be cloned. 1`] = `
Object {
  "value": "foo",
}
`;

exports[`FilePathParameter Can be saved to JSON (serialization). 1`] = `"{\\"value\\":\\"foo\\"}"`;
